### PR TITLE
Document all the grpc api and generate html documentation

### PR DESCRIPTION
**Pull request by gianarb**
## Description

Document the grpc api and generate an HTML page that can be deployed somewhere

- [x] bootstrap doc
- [x] find a way to generate an HTML version of it
- [x] everybody @thebsdbox , @mmlb , @nathangoulding @kdeng3849 please please please, send PR against this one with more documentation, some of the fields and methods need love.

## Why is this needed

Documentation is important.

Fixes: #219 probably many more

## How Has This Been Tested?

`make grpc/gen_doc` and looking at the generated HTML page in `doc` folder

### CODE REVIEW

File: protos/hardware/hardware.proto [hunk 1]
```diff
@@ -1,3 +1,7 @@
+/*
+ * Hardware represents a single device managed by Tinkerbell.
+ * It can be identifed with its own UUID but more in general it can be everything that has a Mac address.
```
Comment (thebsdbox) [inline protos/hardware/hardware.proto:3]: ~everything~ anything that has a ~Mac~ MAC address.

File: protos/hardware/hardware.proto [hunk 2]
```diff
@@ -65,63 +71,162 @@ service HardwareService {
   };
 }
 
+/*
+ * PushRequest is the body for the Push method. It contains information about
+ * an hardware.
```
Comment (thebsdbox) [inline protos/hardware/hardware.proto:76]: ~an~ a piece of hardware

File: protos/hardware/hardware.proto [hunk 3]
```diff
@@ -65,63 +71,162 @@ service HardwareService {
...
+ */
 message PushRequest {
+  /*
+   * hardware describes how the hardware you want to register to Tinkerbell looks like.
```
Comment (thebsdbox) [inline protos/hardware/hardware.proto:80]: ~h~Hardware describes ~how~ the hardware ..

Comment (thebsdbox) [reply protos/hardware/hardware.proto:80]: to Tinkerbell "and what it" looks like.

File: protos/hardware/hardware.proto [hunk 4]
```diff
@@ -65,63 +71,162 @@ service HardwareService {
...
+   * Hostname, mac address, DHCP, network interfaces, metadata and so on.
```
Comment (thebsdbox) [inline protos/hardware/hardware.proto:81]: ~mac~MAC

File: protos/hardware/hardware.proto [hunk 5]
```diff
@@ -65,63 +71,162 @@ service HardwareService {
...
+      /*
+       * Netboot pilots netbooting. You can select Osie, the in memory
+       * operating system used during netbooting.
+       */
       Netboot netboot = 2;
     }
     reserved 1; // obsolete dhcp
     reserved 2; // obsolete netboot
+    /*
+     * Configure the network interfaces you have in your hardware and how their
```
Comment (thebsdbox) [inline protos/hardware/hardware.proto:227]: and ~how~ defines their

File: protos/workflow/workflow.proto [hunk 6]
```diff
@@ -7,33 +11,60 @@ package github.com.tinkerbell.tink.protos.workflow;
 import "google/api/annotations.proto";
 import "google/protobuf/timestamp.proto";
 
+/*
+ * WorkflowSerive exposese various capabilities when it comes to starting and
```
Comment (thebsdbox) [inline protos/workflow/workflow.proto:15]: ~WorkflowSerive~ WorkflowService ~exposeses~ exposes

File: protos/workflow/workflow.proto [hunk 7]
```diff
@@ -7,33 +11,60 @@ package github.com.tinkerbell.tink.protos.workflow;
...
+ * monitoring a workflow
+ */
 service WorkflowService {
+  /*
+   * CreateWorkflow targets a specific hardware and it starts from a particular
+   * template. From now one the selected hardware is capable of picking the
```
Comment (thebsdbox) [inline protos/workflow/workflow.proto:21]: From now on~e~ the selected hardware

**Review by thebsdbox — CHANGES_REQUESTED** [review]
There are a few things that I noticed, but I think this is a great start to having documentation around the API.

File: protos/hardware/hardware.proto [hunk 8]
```diff
@@ -65,63 +71,162 @@ service HardwareService {
...
       int64 family = 4;
     }
+    /*
+     * The mac address for your hardware.  It is important to get it
+     * right because it is used from Tinkerbell to dispatch workflows to the
+     * right destination.
+     */
     string mac = 1;
-    reserved 2; // obsolete ip
+    reserved 2;
```
Comment (kdeng3849) [inline protos/hardware/hardware.proto:122]: can we keep the `// obsolete` tombstones just so we have a clear record of what was deprecated / reason why the number is reserved?

File: protos/workflow/workflow.proto [hunk 9]
```diff
@@ -50,101 +81,280 @@ service WorkflowService {
...
   repeated string command = 5;
-  repeated string on_timeout = 6;
-  repeated string on_failure = 7;
+  /*
+   * On timeout used to be a way to execute an action if the current one timesout
+   * but with the event system in place this is not needed anymore
+   */
+  repeated string on_timeout = 6 [deprecated = true];
+  /*
+   * On failure used to be a way to execute an action if the current one timesout
```
Comment (kdeng3849) [inline protos/workflow/workflow.proto:314]: `timesout` should be `times out`

**Review by kdeng3849 — DISMISSED** [review]

**Review by thebsdbox — APPROVED** [review]

**Review by mmlb — APPROVED** [review]

**Pull request merged**